modules/core/src/main/java/org/apache/ignite/internal/processors/datastructures/GridCacheSemaphoreImpl.java

```diff
@@ -662,6 +667,31 @@ private void initializeSemaphore() throws IgniteCheckedException {
         }
     }
 
+    /** {@inheritDoc} */
+    @Override public <T> IgniteFuture<T> acquireAndExecute(IgniteCallable<T> callable,
+                                                           int numPermits) throws Exception {
+        acquire(numPermits);
+
+        T value = callable.call();
```

Comment:
The method signature implies that it's asynchronous. The implementation, however, executes the provided callable synchronously. It should be submitted to the local public thread pool instead.